cmd/juju/application/deploy.go

```diff
@@ -640,7 +643,52 @@ func (c *DeployCommand) deployBundle(
 	channel params.Channel,
 	apiRoot DeployAPI,
 	bundleStorage map[string]map[string]storage.Constraints,
-) error {
+) (rErr error) {
+	bakeryClient, err := c.BakeryClient()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	modelUUID, ok := apiRoot.ModelUUID()
+	if !ok {
+		return errors.New("API connection is controller-only (should never happen)")
+	}
+
+	for application, applicationSpec := range data.Applications {
+		if applicationSpec.Plan != "" {
+			for _, step := range c.Steps {
+				s := step
+				registrationStep, ok := step.(*RegisterMeteredCharm)
+				if ok && applicationSpec.Plan != "default" {
+					registrationStep.Plan = applicationSpec.Plan
+					s = registrationStep
+				}
+
+				charmURL, err := charm.ParseURL(applicationSpec.Charm)
+				if err != nil {
+					return errors.Trace(err)
+				}
+
+				deployInfo := DeploymentInfo{
+					CharmID:         charmstore.CharmID{URL: charmURL},
+					ApplicationName: application,
+					ModelUUID:       modelUUID,
+				}
+
+				err = s.RunPreBundle(apiRoot, bakeryClient, ctx, deployInfo)
```

Comment:
Here I'd just look up the plan if "default" with `getDefaultPlan`, then call `registerMetrics`, failing immediately on any error. If the bundle author specified a plan, the intent is to deploy as specified, no magic.